File: src/Collega.Application/Fields/FieldDefinitionService.cs
using System.Text.Json;
using Collega.Application.Abstractions;
using Collega.Application.Exceptions;
using Collega.Domain.Auditing;
using Collega.Domain.Enums;
using Collega.Domain.Fields;

namespace Collega.Application.Fields;

/// <summary>
/// User-Defined Field definition use cases (SPEC/20-feature-user-defined-fields.md). Site Admin may
/// manage any organization; Org Admin only their own. Listing/getting active definitions is available
/// to any authenticated member of the organization because every idea form needs the field schema;
/// create/update/delete/reorder and <c>includeDeleted</c> are admin-only.
/// </summary>
public sealed class FieldDefinitionService : IFieldDefinitionService
{
    private const int DisplayOrderStep = 10;

    private readonly IFieldDefinitionRepository _repository;
    private readonly IOrganizationRepository _organizationRepository;
    private readonly IUnitOfWork _unitOfWork;
    private readonly IAuditEventWriter _auditEventWriter;
    private readonly ICurrentUserContext _currentUser;
    private readonly IClock _clock;

    public FieldDefinitionService(
        IFieldDefinitionRepository repository,
        IOrganizationRepository organizationRepository,
        IUnitOfWork unitOfWork,
        IAuditEventWriter auditEventWriter,
        ICurrentUserContext currentUser,
        IClock clock)
    {
        _repository = repository;
        _organizationRepository = organizationRepository;
        _unitOfWork = unitOfWork;
        _auditEventWriter = auditEventWriter;
        _currentUser = currentUser;
        _clock = clock;
    }

    public async Task<IReadOnlyList<FieldDefinitionModel>> ListAsync(Guid organizationId, bool includeDeleted, CancellationToken cancellationToken = default)
    {
        EnsureReadScope(organizationId);
        await EnsureOrganizationExistsAsync(organizationId, cancellationToken);

        // Archived definitions are visible to admins only; members always see the active schema only.
        var effectiveIncludeDeleted = includeDeleted && IsAdminScope(organizationId);

        var definitions = await _repository.ListByOrganizationAsync(organizationId, effectiveIncludeDeleted, cancellationToken);
        return definitions.Select(ToModel).ToList();
    }

    public async Task<FieldDefinitionModel> GetAsync(Guid organizationId, Guid id, CancellationToken cancellationToken = default)
    {
        EnsureReadScope(organizationId);

        var definition = await _repository.GetByIdAsync(id, cancellationToken);
        if (definition is null || definition.OrganizationId != organizationId)
        {
            throw new NotFoundAppException("Field definition not found.");
        }

        // Members cannot see archived definitions; admins can.
        if (definition.IsDeleted && !IsAdminScope(organizationId))
        {
            throw new NotFoundAppException("Field definition not found.");
        }

        return ToModel(definition);
    }

    public async Task<FieldDefinitionModel> CreateAsync(Guid organizationId, CreateFieldDefinitionCommand command, CancellationToken cancellationToken = default)
    {
        EnsureAdminScope(organizationId);
        await EnsureOrganizationExistsAsync(organizationId, cancellationToken);

        var fieldType = ParseFieldType(command.FieldType);
        await EnsureNameAvailableAsync(organizationId, command.Name, excludeId: null, cancellationToken);

        var now = _clock.UtcNow;
        var displayOrder = command.DisplayOrder ?? await NextDisplayOrderAsync(organizationId, cancellationToken);
        var options = ToOptionInputs(command.Options);

        var definition = AsValidation(() => FieldDefinition.Create(
            organizationId, command.Name, command.Description, fieldType, command.IsRequired, displayOrder, options, now, _currentUser.UserId));

        await _repository.AddAsync(definition, cancellationToken);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("FieldDefinitionCreated", organizationId, definition.Id,
            $"Field '{definition.Name}' created.", now,
            new { definition.Name, FieldType = definition.FieldType.ToString(), definition.IsRequired }, cancellationToken);

        return ToModel(definition);
    }

    public async Task<FieldDefinitionModel> UpdateAsync(Guid organizationId, Guid id, UpdateFieldDefinitionCommand command, CancellationToken cancellationToken = default)
    {
        EnsureAdminScope(organizationId);

        var definition = await _repository.GetByIdAsync(id, cancellationToken);
        if (definition is null || definition.OrganizationId != organizationId || definition.IsDeleted)
        {
            throw new NotFoundAppException("Field definition not found.");
        }

        var fieldType = ParseFieldType(command.FieldType);
        if (fieldType != definition.FieldType)
        {
            throw new ValidationAppException("fieldType", new[] { "Field type cannot be changed after creation." });
        }

        await EnsureNameAvailableAsync(organizationId, command.Name, excludeId: id, cancellationToken);

        var now = _clock.UtcNow;
        var displayOrder = command.DisplayOrder ?? definition.DisplayOrder;

        AsValidation(() =>
        {
            definition.Update(command.Name, command.Description, command.IsRequired, displayOrder, now, _currentUser.UserId);
            if (definition.IsOptionBacked)
            {
                definition.SetOptions(ToOptionInputs(command.Options), now, _currentUser.UserId);
            }
        });

        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("FieldDefinitionUpdated", organizationId, definition.Id,
            $"Field '{definition.Name}' updated.", now,
            new { definition.Name, definition.IsRequired, definition.DisplayOrder }, cancellationToken);

        return ToModel(definition);
    }

    public async Task DeleteAsync(Guid organizationId, Guid id, CancellationToken cancellationToken = default)
    {
        EnsureAdminScope(organizationId);

        var definition = await _repository.GetByIdAsync(id, cancellationToken);
        if (definition is null || definition.OrganizationId != organizationId)
        {
            throw new NotFoundAppException("Field definition not found.");
        }

        if (definition.IsDeleted)
        {
            // Soft delete is idempotent.
            return;
        }

        var now = _clock.UtcNow;
        definition.SoftDelete(now, _currentUser.UserId);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        await AuditAsync("FieldDefinitionDeleted", organizationId, definition.Id,
            $"Field '{definition.Name}' archived.", now, null, cancellationToken);
    }

    public async Task ReorderAsync(Guid organizationId, ReorderFieldDefinitionsCommand command, CancellationToken cancellationToken = default)
    {
        EnsureAdminScope(organizationId);
        await EnsureOrganizationExistsAsync(organizationId, cancellationToken);

        var definitions = await _repository.ListActiveTrackedByOrganizationAsync(organizationId, cancellationToken);
        var byId = definitions.ToDictionary(d => d.Id);

        var now = _clock.UtcNow;
        var order = 0;
        foreach (var definitionId in command.OrderedIds ?? Array.Empty<Guid>())
        {
            if (!byId.TryGetValue(definitionId, out var definition))
            {
                // Unknown or already-archived ids are ignored (the active set is authoritative).
                continue;
            }

            order += DisplayOrderStep;
            definition.SetDisplayOrder(order, now, _currentUser.UserId);
        }

        await _unitOfWork.SaveChangesAsync(cancellationToken);
        await AuditAsync("FieldDefinitionsReordered", organizationId, null, "Field display order updated.", now, null, cancellationToken);
    }

    private async Task<int> NextDisplayOrderAsync(Guid organizationId, CancellationToken cancellationToken)
    {
        var existing = await _repository.ListByOrganizationAsync(organizationId, includeDeleted: true, cancellationToken);
        return existing.Count == 0 ? DisplayOrderStep : existing.Max(d => d.DisplayOrder) + DisplayOrderStep;
    }

    private async Task EnsureNameAvailableAsync(Guid organizationId, string name, Guid? excludeId, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            // Let the domain produce the canonical "Name is required." message.
            return;
        }

        if (await _repository.ExistsActiveByNameAsync(organizationId, name, excludeId, cancellationToken))
        {
            throw new ValidationAppException("name", new[] { $"A field named '{name.Trim()}' already exists in this organization." });
        }
    }

    private async Task EnsureOrganizationExistsAsync(Guid organizationId, CancellationToken cancellationToken)
    {
        _ = await _organizationRepository.GetByIdAsync(organizationId, cancellationToken)
            ?? throw new NotFoundAppException("Organization not found.");
    }

    private static FieldType ParseFieldType(string fieldType)
    {
        if (!string.IsNullOrWhiteSpace(fieldType)
            && Enum.TryParse<FieldType>(fieldType.Trim(), ignoreCase: true, out var parsed)
            && Enum.IsDefined(parsed))
        {
            return parsed;
        }

        var supported = string.Join(", ", Enum.GetNames<FieldType>());
        throw new ValidationAppException("fieldType", new[] { $"Field type must be one of: {supported}." });
    }

    private static IReadOnlyList<FieldOptionInput> ToOptionInputs(IReadOnlyList<FieldOptionCommand>? options) =>
        (options ?? Array.Empty<FieldOptionCommand>())
            .Select((o, index) => new FieldOptionInput(o.OptionId, o.Label, o.DisplayOrder == 0 ? index : o.DisplayOrder))
            .ToList();

    /// <summary>Site Admin (any org) or a member of the target organization (any role) may read the schema.</summary>
    private void EnsureReadScope(Guid organizationId)
    {
        var role = RequireAuthenticatedRole();
        if (role == Role.SiteAdmin || _currentUser.OrganizationId == organizationId)
        {
            return;
        }

        throw new NotFoundAppException("Organization not found.");
    }

    /// <summary>Site Admin (any org) or Org Admin (own org) may manage field definitions.</summary>
    private void EnsureAdminScope(Guid organizationId)
    {
        var role = RequireAuthenticatedRole();
        if (role == Role.SiteAdmin)
        {
            return;
        }

        if (role == Role.OrgAdmin)
        {
            if (_currentUser.OrganizationId == organizationId)
            {
                return;
            }

            throw new NotFoundAppException("Organization not found.");
        }

        throw new ForbiddenAppException("You are not allowed to manage field definitions in this organization.");
    }

    private bool IsAdminScope(Guid organizationId)
    {
        if (!_currentUser.IsAuthenticated || _currentUser.Role is null)
        {
            return false;
        }

        return _currentUser.Role == Role.SiteAdmin
            || (_currentUser.Role == Role.OrgAdmin && _currentUser.OrganizationId == organizationId);
    }

    private Role RequireAuthenticatedRole()
    {
        if (!_currentUser.IsAuthenticated || _currentUser.Role is null)
        {
            throw new UnauthorizedAppException("Caller identity could not be resolved.");
        }

        return _currentUser.Role.Value;
    }

    private static FieldDefinitionModel ToModel(FieldDefinition d) => new(
        d.Id,
        d.OrganizationId,
        d.Name,
        d.Description,
        d.FieldType.ToString(),
        d.IsRequired,
        d.DisplayOrder,
        d.IsDeleted,
        d.Options
            .OrderBy(o => o.DisplayOrder)
            .Select(o => new FieldOptionModel(o.Id, o.Label, o.DisplayOrder))
            .ToList());

    private static T AsValidation<T>(Func<T> action)
    {
        try
        {
            return action();
        }
        catch (Exception ex) when (ex is ArgumentException or InvalidOperationException)
        {
            throw new ValidationAppException("fieldDefinition", new[] { ex.Message });
        }
    }

    private static void AsValidation(Action action) =>
        AsValidation<object?>(() =>
        {
            action();
            return null;
        });

    private async Task AuditAsync(
        string eventType,
        Guid organizationId,
        Guid? entityId,
        string message,
        DateTime nowUtc,
        object? metadata,
        CancellationToken cancellationToken)
    {
        var metadataJson = metadata is null ? null : JsonSerializer.Serialize(metadata);
        var auditEvent = AuditEvent.Create(eventType, "FieldDefinition", message, nowUtc, organizationId, _currentUser.UserId, entityId, metadataJson);
        await _auditEventWriter.WriteAsync(auditEvent, cancellationToken);
    }
}

File: packages/application/src/fields/field-definition-service.ts
// User-Defined Field definition use cases (SPEC/20-feature-user-defined-fields.md). Site Admin
// may manage any organization; Org Admin only their own. Listing/getting active definitions is
// available to any authenticated member of the organization because every idea form needs the
// field schema; create/update/delete/reorder and `includeDeleted` are admin-only.

import { randomUUID } from 'node:crypto'
import { FieldType, Role } from '@collega/domain/enums'
import {
  createFieldDefinition,
  type FieldDefinition,
  FieldDefinitionDomainError,
  type FieldOptionInput,
  isOptionBackedFieldType,
  setFieldDefinitionDisplayOrder,
  setFieldDefinitionOptions,
  softDeleteFieldDefinition,
  updateFieldDefinition,
} from '@collega/domain/fields'
import {
  type AuditEventWriter,
  attributeAudit,
  type Clock,
  type CurrentUserContext,
  ensureNotDirectSiteAdmin,
  ForbiddenError,
  NotFoundError,
  UnauthorizedError,
  type UnitOfWork,
  ValidationError,
} from '../common/index.js'
import type {
  CreateFieldDefinitionCommand,
  FieldDefinitionModel,
  FieldOptionCommand,
  ReorderFieldDefinitionsCommand,
  UpdateFieldDefinitionCommand,
} from './models.js'
import type { FieldDefinitionRepository, OrganizationExistenceLookup } from './ports.js'

const DISPLAY_ORDER_STEP = 10

export class FieldDefinitionService {
  constructor(
    private readonly fieldDefinitions: FieldDefinitionRepository,
    private readonly organizations: OrganizationExistenceLookup,
    private readonly unitOfWork: UnitOfWork,
    private readonly auditEvents: AuditEventWriter,
    private readonly currentUser: CurrentUserContext,
    private readonly clock: Clock,
  ) {}

  async list(
    organizationId: string,
    includeDeleted: boolean,
  ): Promise<readonly FieldDefinitionModel[]> {
    this.ensureReadScope(organizationId)
    await this.ensureOrganizationExists(organizationId)

    // Archived definitions are visible to admins only; members always see the active schema only.
    const effectiveIncludeDeleted = includeDeleted && this.isAdminScope(organizationId)
    const definitions = await this.fieldDefinitions.listByOrganization(
      organizationId,
      effectiveIncludeDeleted,
    )
    return definitions.map(toModel)
  }

  async getById(organizationId: string, id: string): Promise<FieldDefinitionModel> {
    this.ensureReadScope(organizationId)

    const definition = await this.fieldDefinitions.getById(id)
    if (definition === null || definition.organizationId !== organizationId) {
      throw new NotFoundError('Field definition not found.')
    }

    // Members cannot see archived definitions; admins can.
    if (definition.isDeleted && !this.isAdminScope(organizationId)) {
      throw new NotFoundError('Field definition not found.')
    }

    return toModel(definition)
  }

  async create(
    organizationId: string,
    command: CreateFieldDefinitionCommand,
  ): Promise<FieldDefinitionModel> {
    this.ensureAdminScope(organizationId)
    await this.ensureOrganizationExists(organizationId)

    const fieldType = parseFieldType(command.fieldType)
    await this.ensureNameAvailable(organizationId, command.name, null)

    const now = this.clock.now()
    const displayOrder = command.displayOrder ?? (await this.nextDisplayOrder(organizationId))
    const options = toOptionInputs(command.options)

    const definition = runDomain(createFieldDefinition, {
      id: randomUUID(),
      organizationId,
      name: command.name,
      description: command.description,
      fieldType,
      isRequired: command.isRequired,
      displayOrder,
      options,
      nowUtc: now,
      actorUserId: this.currentUser.userId,
    })

    await this.fieldDefinitions.add(definition)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'FieldDefinitionCreated',
      organizationId,
      definition.id,
      `Field '${definition.name}' created.`,
      now,
      { name: definition.name, fieldType: definition.fieldType, isRequired: definition.isRequired },
    )

    return toModel(definition)
  }

  async update(
    organizationId: string,
    id: string,
    command: UpdateFieldDefinitionCommand,
  ): Promise<FieldDefinitionModel> {
    this.ensureAdminScope(organizationId)

    const existing = await this.fieldDefinitions.getById(id)
    if (existing === null || existing.organizationId !== organizationId || existing.isDeleted) {
      throw new NotFoundError('Field definition not found.')
    }

    const fieldType = parseFieldType(command.fieldType)
    if (fieldType !== existing.fieldType) {
      throw new ValidationError('One or more fields are invalid.', {
        fieldType: ['Field type cannot be changed after creation.'],
      })
    }

    await this.ensureNameAvailable(organizationId, command.name, id)

    const now = this.clock.now()
    const displayOrder = command.displayOrder ?? existing.displayOrder

    let definition = runDomain(
      updateFieldDefinition,
      existing,
      {
        name: command.name,
        description: command.description,
        isRequired: command.isRequired,
        displayOrder,
      },
      now,
      this.currentUser.userId,
    )
    if (isOptionBackedFieldType(definition.fieldType)) {
      definition = runDomain(
        setFieldDefinitionOptions,
        definition,
        toOptionInputs(command.options),
        now,
        this.currentUser.userId,
      )
    }

    await this.fieldDefinitions.save(definition)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'FieldDefinitionUpdated',
      organizationId,
      definition.id,
      `Field '${definition.name}' updated.`,
      now,
      {
        name: definition.name,
        isRequired: definition.isRequired,
        displayOrder: definition.displayOrder,
      },
    )

    return toModel(definition)
  }

  async delete(organizationId: string, id: string): Promise<void> {
    this.ensureAdminScope(organizationId)

    const existing = await this.fieldDefinitions.getById(id)
    if (existing === null || existing.organizationId !== organizationId) {
      throw new NotFoundError('Field definition not found.')
    }

    if (existing.isDeleted) {
      // Soft delete is idempotent.
      return
    }

    const now = this.clock.now()
    const definition = runDomain(softDeleteFieldDefinition, existing, now, this.currentUser.userId)
    await this.fieldDefinitions.save(definition)
    await this.unitOfWork.saveChanges()

    await this.audit(
      'FieldDefinitionDeleted',
      organizationId,
      definition.id,
      `Field '${definition.name}' archived.`,
      now,
      null,
    )
  }

  /**
   * Replaces every active definition's display order atomically - every `save` below is staged
   * and committed with a single `unitOfWork.saveChanges()` after the loop, mirroring the .NET's
   * single post-loop `SaveChangesAsync` (SPEC/decisions.md 2026-09-06 "Wave B conventions").
   */
  async reorder(organizationId: string, command: ReorderFieldDefinitionsCommand): Promise<void> {
    this.ensureAdminScope(organizationId)
    await this.ensureOrganizationExists(organizationId)

    const definitions = await this.fieldDefinitions.listActiveByOrganization(organizationId)
    const byId = new Map(definitions.map((d) => [d.id, d] as const))

    const now = this.clock.now()
    let order = 0
    const assigned = new Set<string>()
    for (const definitionId of command.orderedIds ?? []) {
      // Unknown/archived ids and repeats are ignored (the active set is authoritative).
      const definition = byId.get(definitionId)
      if (definition === undefined || assigned.has(definitionId)) {
        continue
      }
      assigned.add(definitionId)

      order += DISPLAY_ORDER_STEP
      const updated = setFieldDefinitionDisplayOrder(
        definition,
        order,
        now,
        this.currentUser.userId,
      )
      await this.fieldDefinitions.save(updated)
    }

    // Any active definitions the caller omitted keep their prior relative order, appended after
    // the explicitly ordered ones - so every active field gets a fresh, collision-free display
    // order (`definitions` is already sorted by display order).
    for (const definition of definitions) {
      if (assigned.has(definition.id)) {
        continue
      }
      order += DISPLAY_ORDER_STEP
      const updated = setFieldDefinitionDisplayOrder(
        definition,
        order,
        now,
        this.currentUser.userId,
      )
      await this.fieldDefinitions.save(updated)
    }

    await this.unitOfWork.saveChanges()
    await this.audit(
      'FieldDefinitionsReordered',
      organizationId,
      null,
      'Field display order updated.',
      now,
      null,
    )
  }

  private async nextDisplayOrder(organizationId: string): Promise<number> {
    const existing = await this.fieldDefinitions.listByOrganization(organizationId, true)
    if (existing.length === 0) {
      return DISPLAY_ORDER_STEP
    }
    return Math.max(...existing.map((d) => d.displayOrder)) + DISPLAY_ORDER_STEP
  }

  private async ensureNameAvailable(
    organizationId: string,
    name: string,
    excludeId: string | null,
  ): Promise<void> {
    if (!name || name.trim().length === 0) {
      // Let the domain produce the canonical "Name is required." message.
      return
    }

    if (await this.fieldDefinitions.existsActiveByName(organizationId, name, excludeId)) {
      throw new ValidationError('One or more fields are invalid.', {
        name: [`A field named '${name.trim()}' already exists in this organization.`],
      })
    }
  }

  private async ensureOrganizationExists(organizationId: string): Promise<void> {
    if (!(await this.organizations.existsById(organizationId))) {
      throw new NotFoundError('Organization not found.')
    }
  }

  /** Site Admin (any org) or a member of the target organization (any role) may read the schema. */
  private ensureReadScope(organizationId: string): void {
    const role = this.requireAuthenticatedRole()
    if (role === Role.SiteAdmin || this.currentUser.organizationId === organizationId) {
      return
    }
    throw new NotFoundError('Organization not found.')
  }

  /** Site Admin (any org) or Org Admin (own org) may manage field definitions. */
  private ensureAdminScope(organizationId: string): void {
    const role = this.requireAuthenticatedRole()

    // Rule 25: a Site Admin acting as themselves may not mutate org-owned content - View As is
    // the path. No impersonation branch is needed: while a session is live this role is the
    // target's, so the guard does not fire. Read paths use ensureReadScope and are unaffected.
    ensureNotDirectSiteAdmin(this.currentUser)

    // No SiteAdmin pass-through below, deliberately: the guard above has already refused that
    // role, so a `return` here would be dead code that reads like a live bypass.
    if (role === Role.OrgAdmin) {
      if (this.currentUser.organizationId === organizationId) {
        return
      }
      throw new NotFoundError('Organization not found.')
    }

    throw new ForbiddenError(
      'You are not allowed to manage field definitions in this organization.',
    )
  }

  private isAdminScope(organizationId: string): boolean {
    if (!this.currentUser.isAuthenticated || this.currentUser.role === null) {
      return false
    }
    return (
      this.currentUser.role === Role.SiteAdmin ||
      (this.currentUser.role === Role.OrgAdmin &&
        this.currentUser.organizationId === organizationId)
    )
  }

  private requireAuthenticatedRole(): Role {
    if (!this.currentUser.isAuthenticated || this.currentUser.role === null) {
      throw new UnauthorizedError('Caller identity could not be resolved.')
    }
    return this.currentUser.role
  }

  private async audit(
    eventType: string,
    organizationId: string,
    entityId: string | null,
    message: string,
    occurredAtUtc: Date,
    metadata: Record<string, unknown> | null,
  ): Promise<void> {
    // Rule 14: while acting as someone, the real administrator is the actor and the target moves
    // to onBehalfOfUserId - an audit row must never read as though the target did it.
    const attribution = attributeAudit(this.currentUser, this.currentUser.userId)
    await this.auditEvents.write({
      eventType,
      entityType: 'FieldDefinition',
      message,
      occurredAtUtc,
      organizationId,
      attribution,
      entityId,
      metadataJson: metadata === null ? null : JSON.stringify(metadata),
    })
  }
}

function toModel(definition: FieldDefinition): FieldDefinitionModel {
  return {
    fieldDefinitionId: definition.id,
    organizationId: definition.organizationId,
    name: definition.name,
    description: definition.description,
    fieldType: definition.fieldType,
    isRequired: definition.isRequired,
    displayOrder: definition.displayOrder,
    isDeleted: definition.isDeleted,
    options: [...definition.options]
      .sort((a, b) => a.displayOrder - b.displayOrder)
      .map((o) => ({ optionId: o.id, label: o.label, displayOrder: o.displayOrder })),
  }
}

function toOptionInputs(
  options: readonly FieldOptionCommand[] | null | undefined,
): readonly FieldOptionInput[] {
  return (options ?? []).map((o, index) => ({
    id: o.optionId ?? randomUUID(),
    label: o.label,
    displayOrder: o.displayOrder === 0 ? index : o.displayOrder,
  }))
}

function parseFieldType(fieldType: string): FieldType {
  const trimmed = (fieldType ?? '').trim()
  const match = Object.values(FieldType).find(
    (value) => value.toLowerCase() === trimmed.toLowerCase(),
  )
  if (trimmed.length > 0 && match !== undefined) {
    return match
  }

  throw new ValidationError('One or more fields are invalid.', {
    fieldType: [`Field type must be one of: ${Object.values(FieldType).join(', ')}.`],
  })
}

/**
 * Wraps a domain transition so a `FieldDefinitionDomainError` surfaces as a `ValidationError`
 * bucketed under the fixed `fieldDefinition` key - mirroring the .NET `FieldDefinitionService`'s
 * shared `AsValidation` helper exactly, INCLUDING the fact that it does not key by the specific
 * property that failed. This is a deliberate departure from `IdeaDomainError`/
 * `BoardInvariantError`'s per-field keying (the `runDomain` reference elsewhere in Wave B): every
 * `FieldDefinitionService` call site in the .NET used one shared `AsValidation` that always threw
 * `ValidationAppException("fieldDefinition", new[] { ex.Message })` regardless of which invariant
 * fired, and the corpus records whatever that produced.
 */
function runDomain<Args extends readonly unknown[], T>(fn: (...args: Args) => T, ...args: Args): T {
  try {
    return fn(...args)
  } catch (error) {
    if (error instanceof FieldDefinitionDomainError) {
      throw new ValidationError('One or more fields are invalid.', {
        fieldDefinition: [error.message],
      })
    }
    throw error
  }
}
